modules/frontend/retry.go

```diff
@@ -42,25 +50,23 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 
 		resp, err := r.next.Do(req)
 
-		triesLeft--
-		if triesLeft == 0 {
-			return resp, err
-		}
-
+		// do not retry if no error and response is not HTTP 5xx
 		if err == nil && resp.StatusCode/100 != 5 {
 			return resp, nil
 		}
 
-		statusCode := 0
-		if resp != nil {
-			statusCode = resp.StatusCode
+		// do not retry if GRPC error contains response that is not HTTP 5xx
+		httpResp, ok := httpgrpc.HTTPResponseFromError(err)
+		if ok && httpResp.Code/100 != 5 {
+			return resp, err
+		}
+
+		// reached max retries
+		tries++
+		if tries >= r.maxRetries {
+			return resp, err
 		}
 
-		span.LogFields(
-			ot_log.String("msg", "error processing request"),
-			ot_log.Int("status_code", statusCode),
-			ot_log.Error(err),
-			ot_log.Int("triesLeft", triesLeft),
-		)
+		span.LogFields(ot_log.String("msg", "error processing request. retrying"))
```

Comment:
What is the thinking to remove the extra data from the span? It seems helpful.

Reply:
I removed `ot_log.Error(err)` because this triggers https://github.com/grafana/tempo/issues/843
Agree it's valuable info, but for some reason calling `err.Error()` can cause panics 🙃 

I'll re-add the tries.

Reply:
Found a safe way to log the error 🙃